### DISCUSSION TITLE
First-class stream type

### DISCUSSION THREAD

**Discussion by kaathewisegit** [nushell/nushell#16660] [Ideas]
### Related problem

I think there are currently a couple issues with how streams are exposed in Nushell:

- The streams are opaque.  Some commands, such as `lines` does one type of streaming, while `fliter` does another.  Despite this, both commands return a `List` type, and so do non-streaming commands.  This causes problems such as #8016 and #10708.
- To add to the previous point, the fact that `$in` collects the incoming stream is also non-obvious: #9551.
- Performance issues, including #10763, #6174.
- Scripting troubles.  Currently, to process a stream, a command has to start processing right away, without any intermediary setup.  Otherwise one has to use `$in`, which collects.  Related to #9373.

### Describe the solution you'd like

I think a new first-class stream type would be useful.  I am unsure how it can be implemented, however, as it has to work for both strings and lists.

**Comment by fdncred** [comment]
I'm not quite following with your leading bullet.
```nushell
❯ open toolkit.nu | lines | describe
list<string> (stream)
```

```nushell
❯ ls | filter { $in.type == dir } | describe
table<name: string, type: string, size: filesize, modified: date> (stream)
```

However I do see that `lines` has `collect` in it which stops the streaming I believe.
https://github.com/nushell/nushell/blob/15c22db8f49d42cba97dc5f4bc1de24ac8c4587b/crates/nu-command/src/filters/lines.rs#L54
https://github.com/nushell/nushell/blob/15c22db8f49d42cba97dc5f4bc1de24ac8c4587b/crates/nu-command/src/filters/lines.rs#L90
https://github.com/nushell/nushell/blob/15c22db8f49d42cba97dc5f4bc1de24ac8c4587b/crates/nu-command/src/filters/lines.rs#L201


We've talked about creating a streaming Value like we used to have recently.
https://hackmd.io/TZsIBlpyQs-GVVwsdCcmQA#Streaming-as-a-first-class-citizen

I believe we decided we wanted to do this but no one has tried yet. If you'd like to give it a whirl, please say so in #design-discussion on discord. You could also jump into our wednesday meetings if you want to chat about doing it.

**Comment by kaathewisegit** [comment]
Oh, I was unaware there was an ongoing discussion!  This makes this issue somewhat redundant.

As for `lines`, it does some re-streaming for `PipelineData::ExternalStream`, so it works with things like `tail -f foo | lines`.

**Comment by kaathewisegit** [comment]
I think I'll finish with highlighting this month and try making a draft for streams.

**Comment by fdncred** [comment]
Sounds good. It's also fine to leave this issue as a conversation and status place. We'd like to move in that direction.

**Comment by flying-sheep** [comment]
This would be great! Currently I have giant yaml files that are too big to parse, but I know their structure.

It would be great if I could just split the stream before each occurrence of the regex `^-` and put each produced chunk through `from yaml`.

**Comment by 132ikl** [comment]
Since I'm not certain this is the direction we want to go, I'm going to convert this into a discussion. If we decide to definitely go this route we can promote it back into an issue

**Comment by nome** [comment]
Other use cases:
* examining a pipeline result in `$env.config.hooks.display_output` without collecting it, in order to implement access to previous results with some guard rails for large outputs (as recently discussed on Discord by multiple people, see #2731 nushell/nu_scripts#1176 and https://github.com/YPares/monurepo/tree/master/repage)
* allowing multiple input/output streams, see #549
